Frameworks/Foundation/NSURLConnection.mm

```diff
@@ -19,168 +20,289 @@
 #import "Foundation/NSMutableData.h"
 #import "Foundation/NSError.h"
 #import "Foundation/NSRunLoop.h"
-#import "NSURLConnectionState.h"
 #import "Foundation/NSURLConnection.h"
 #import "LoggingNative.h"
 #import "NSURLProtocolInternal.h"
 
+#import <vector>
+
 static const wchar_t* TAG = L"NSURLConnection";
 
-@implementation NSURLConnection
+// Private helper function that returns a NSURLProtocol object capable of handling a NSURLRequest
+// Returns nil if one cannot be created
+static NSURLProtocol* _protocolForRequest(NSURLRequest* request, id<NSURLProtocolClient> client) {
+    Class cls = [NSURLProtocol _URLProtocolClassForRequest:request];
+    if (!cls || ![cls canInitWithRequest:request]) {
+        return nil;
+    }
 
-/**
- @Status Interoperable
-*/
-+ (BOOL)canHandleRequest:(id)request {
-    return ([NSURLProtocol _URLProtocolClassForRequest:request] != nil) ? YES : NO;
+    return
+        [[[cls alloc] initWithRequest:request cachedResponse:[[NSURLCache sharedURLCache] cachedResponseForRequest:request] client:client]
+            autorelease];
 }
 
-/**
- @Status Caveat
- @Notes queue parameter not supported
-*/
-+ (void)sendAsynchronousRequest:(id)request
-                          queue:(id)queue
-              completionHandler:(void (^)(NSURLResponse*, NSData*, NSError*))completionHandler {
-    TraceVerbose(TAG, L"sendAsynchronousRequest not fully supported");
+// Private class that acts as a helper for sendAsynchronousRequest: (and by extension, sendSynchronousRequest:)
+// Forwards response, data, and error to a completion handler, which is executed on a specified queue.
+@interface _NSURLConnection_DataDelegate : NSObject <NSURLConnectionDataDelegate>
++ (_NSURLConnection_DataDelegate*)delegateWithQueue:(NSOperationQueue*)queue handler:(void (^)(NSURLResponse*, NSData*, NSError*))handler;
+@end
 
-    id response, error;
-    id data = [self sendSynchronousRequest:request returningResponse:&response error:&error];
+@implementation _NSURLConnection_DataDelegate {
+@private
+    StrongId<NSOperationQueue> _operationQueue;
+    StrongId<void (^)(NSURLResponse*, NSData*, NSError*)> _completionHandler;
 
-    completionHandler(response, data, error);
+    StrongId<NSURLResponse> _response;
+    StrongId<NSMutableData> _data;
 }
 
-/**
- @Status Caveat
- @Notes NSError returned is not detailed
-*/
-+ (NSData*)sendSynchronousRequest:(id)request returningResponse:(NSURLResponse**)responsep error:(NSError**)errorp {
-    NSURLConnectionState* state = [[[NSURLConnectionState alloc] init] autorelease];
-    NSURLConnection* connection = [[self alloc] initWithRequest:request delegate:state startImmediately:FALSE];
++ (_NSURLConnection_DataDelegate*)delegateWithQueue:(NSOperationQueue*)queue handler:(void (^)(NSURLResponse*, NSData*, NSError*))handler {
```

Comment:
nit: prefer all factory functions X to go through an appropriate initializer `initWithX:`